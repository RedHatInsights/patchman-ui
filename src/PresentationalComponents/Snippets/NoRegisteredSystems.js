import React from 'react';
import propTypes from 'prop-types';
import {
    Button,
    EmptyState,
    EmptyStateBody,
    EmptyStateIcon,
    EmptyStateHeader,
    EmptyStateFooter
} from '@patternfly/react-core';
import { DisconnectedIcon } from '@patternfly/react-icons';

const NoRegisteredSystems = ({ titleText, bodyText, buttonText }) => (
    <EmptyState>
        <EmptyStateHeader
            titleText={titleText}
            icon={<EmptyStateIcon icon={DisconnectedIcon} />}
            headingLevel="h5"
        />
        <EmptyStateBody>{bodyText}</EmptyStateBody>
        <EmptyStateFooter>
            <Button
                variant="primary"
                component="a"
                href="https://access.redhat.com/products/red-hat-insights#getstarted"
                target="_blank"
                rel="noopener noreferrer"
                className="pf-v5-u-mt-lg"
            >
                {buttonText}
            </Button>
        </EmptyStateFooter>
    </EmptyState>
);

NoRegisteredSystems.propTypes = {
    titleText: propTypes.string,
    bodyText: propTypes.string,
    buttonText: propTypes.string
};

NoRegisteredSystems.defaultProps = {
    titleText: 'Do more with your Red Hat Enterprise Linux environment',
    bodyText: `Connect your systems to keep your Red Hat environment running efficiently,
     with security and compliance with various standards.`,
    buttonText: 'Learn more about connecting your systems'
};
export default NoRegisteredSystems;
