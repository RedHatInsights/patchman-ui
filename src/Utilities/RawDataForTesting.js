// raw data for testing purposes only

export const advisoryRows = [{
    attributes: {
        advisory_type: 1,
        applicable_systems: 1,
        description: 'The microcode_ctl ',
        public_date: '2020-06-24T17:22:25Z',
        synopsis: 'microcode_ctl',
        severity: 3
    },
    id: 'RHEA-2020:2743',
    type: 'advisory'
}];

export const systemAdvisoryRows = [{
    attributes: {
        advisory_type: 3,
        description: 'Kernel-based Virtual.',
        public_date: '2020-06-30T13:38:53Z',
        severity: 3,
        synopsis: 'Important'
    },
    id: 'RHSA-2020:2774',
    type: 'advisory'
}];

export const systemRows = [{
    attributes: {
        display_name: 'automation_host',
        last_evaluation: '2020-08-12T13:57:54.028883Z',
        last_upload: '2020-08-12T09:26:33.891907Z',
        packages_installed: 0,
        packages_updatable: 0,
        rhba_count: 213,
        rhea_count: 35,
        rhsa_count: 77,
        stale: false
    },
    id: '8ddb54f5-aeeb-49b1-8448-c29049d686c1',
    type: 'system'
}];

export const advisoryHeader = {
    attributes: {
        advisory_type: 3,
        description: 'Kernel-based Virtual.',
        public_date: '2020-06-30T13:38:53Z',
        modified_date: '2020-06-30T13:38:53Z',
        severity: 3,
        synopsis: 'Important'
    },
    id: 'RHSA-2020:2774',
    type: 'advisory'
};

export const advisoryDetailRows = {
    data: {
        attributes: {
            cves: ['CVE-2020-11080'],
            description: 'libnghttp2.',
            fixes: null,
            modified_date: '2020-06-25T16:31:29Z',
            packages: { libnghttp2: '1.33.0-3.el8_2.1.x86_64' },
            public_date: '2020-06-25T16:31:29Z',
            references: [],
            severity: 3,
            solution: 'For details',
            synopsis: 'Important: ',
            topic: 'An update .'
        }
    },
    id: 'RHSA-2020:2755',
    type: 'advisory'
};
export const systemPackages = [{
    description: 'This package contains testing data',
    evra: '2.2.*',
    name: 'acl',
    summary: 'Access control list utilities',
    updates: null
}];

export const entityDetail = {
    loaded: true,
    activeApps: [],
    entity: {
        account: '60'
    }
};
