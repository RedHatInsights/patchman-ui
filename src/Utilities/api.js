import { encodeApiParams } from './Helpers';

export function createApiCall(
    endpoint,
    method,
    parameters = undefined,
    data = undefined
) {
    if (parameters && method === 'get') {
        endpoint = endpoint.concat(encodeApiParams(parameters));
    }

    let result = window.insights.chrome.auth
    .getUser()
    .then(() =>
        fetch('/api/patch/v1' + endpoint, {
            method,
            credentials: 'include',
            body: JSON.stringify(data)
        })
    )
    .then(res => {
        if (!res.ok) {
            const contentType = res.headers.get('content-type');
            if (contentType.indexOf('json') !== -1) {
                throw res;
            } else {
                throw {
                    errors: [
                        { status: res.status, detail: res.statusText }
                    ]
                };
            }
        }

        return res.json();
    })
    .catch(caughtError => {
        const error = Promise.resolve(caughtError || {});
        const genericError = {
            title:
                    'There was an error getting data'
        };
        return error.then(error => {
            const [info] = error.errors;
            const result = error.errors && { ...genericError, detail: info.detail, status: info.status } || genericError;
            throw result ;
        });
    });
    return result;
}

export const fetchApplicableAdvisoriesApi = params => {
    return createApiCall('/advisories', 'get', params);
};

export const fetchApplicableSystemAdvisoriesApi = params => {
    let { id, ...allParams } = params;
    return createApiCall(`/systems/${id}/advisories`, 'get', allParams);
};

export const fetchSystems = params => {
    return createApiCall('/systems', 'get', params);
};

export const fetchAdvisoryDetailsApi = params => {
    return createApiCall(`/advisories/${params.advisoryName}`, 'get');
};

export const fetchApplicablePackagesApi = params => {
    let { id, ...allParams } = params;
    return createApiCall(`/systems/${id}/packages`, 'get', allParams);
};

export const fetchAffectedSystems = params => {
    const { id, ...args } = params;
    return createApiCall(`/advisories/${id}/systems`, 'get', args);
};

export const fetchPackagesList = params => {
    return createApiCall('/packages', 'get', params);
};

export const exportAdvisoriesCSV = params => {
    let endpoint = '/export/advisories';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'text/csv' })
    }).then(res => res.text());
};

export const exportAdvisoriesJSON = params => {
    let endpoint = '/export/advisories';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'application/json' })
    }).then(res => res.json());
};

export const exportSystemsCSV = params => {
    let endpoint = '/export/systems';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'text/csv' })
    }).then(res => res.text());
};

export const exportSystemsJSON = params => {
    let endpoint = '/export/systems';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'application/json' })
    }).then(res => res.json());
};
