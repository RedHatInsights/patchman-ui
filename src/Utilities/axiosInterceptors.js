import {
    authInterceptor,
    responseDataInterceptor
} from '@redhat-cloud-services/frontend-components-utilities/files/interceptors';
import axios from 'axios';
import { ReadOnlyNotification } from './constants';
const axiosInstance = axios.create();

export async function  accessCheckInterceptor (config)  {

    const userPermissions = await insights.chrome.getUserPermissions();
    const inventoryPermissionList = ['inventory:*:*', 'inventory:*:read', 'inventory:hosts:read'];

    const hasInventoryAccess = userPermissions.some((access) => inventoryPermissionList.includes(access?.permission));

    if (!hasInventoryAccess) {
        throw new axios.Cancel('Operation canceled. User does not have access to Inventory');
    }

    return config;
}

export function errorInterceptor(err) {

    if (!axios.isCancel(err)) {

        const { response, isAxiosError } = { ...err };

        if (response && isAxiosError) {
            const { status, statusText, data } = response;

            if (!status) {
                throw err;
            } else {
                const genericError = {
                    title:
                            'There was an error getting data'
                };

                const result = { ...genericError, detail: data.error && data.error || statusText, status };

                throw result;
            }
        }

        throw err;
    }
}

export function readOnlyInterceptor(error) {
    if (error.response && error.response.status === 503) {
        const data = ReadOnlyNotification;
        throw data;
    }

    throw error;
}

axiosInstance.interceptors.request.use(accessCheckInterceptor);
axiosInstance.interceptors.request.use(authInterceptor);
axiosInstance.interceptors.response.use(responseDataInterceptor);
axiosInstance.interceptors.response.use(null, readOnlyInterceptor);
axiosInstance.interceptors.response.use(null, errorInterceptor);

export default axiosInstance;
