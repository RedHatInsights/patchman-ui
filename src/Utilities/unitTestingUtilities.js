import { useSelector } from 'react-redux';

/* eslint-disable */
export const initMocks = () => {

    window.insights = {
        loadInventory: (args) => new Promise((resolve, reject) =>  {
            resolve(({
                inventoryConnector: (args) => {
                            const InventoryTable = ({ children }) => <div >A mock passed! {children} </div>;
                            return ({ InventoryTable });
                        },
                        mergeWithEntities: () => {}
            }));
        }),
        loadRemediations: () => new Promise((resolve, reject) =>  {
            resolve({ 
                openWizard: () => new Promise((resolve, reject) =>  { 
                    resolve(true)
                })
            });
        })
    };

};

export const mockStore = (initialState, mutatedState) => {
    const customMiddleWare = store => next => action => {
        useSelector.mockImplementation(callback => {
            return callback(mutatedState);
        });
        next(action);
    };

    const mockStore = configureStore([customMiddleWare]);

    return mockStore(initialState);
}
/* eslint-enable */
