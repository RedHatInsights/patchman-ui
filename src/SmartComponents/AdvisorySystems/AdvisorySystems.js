import { TableVariant } from '@patternfly/react-table';
import { InventoryTable } from '@redhat-cloud-services/frontend-components/Inventory';
import propTypes from 'prop-types';
import React, { useCallback } from 'react';
import { shallowEqual, useDispatch, useSelector, useStore } from 'react-redux';
import { combineReducers } from 'redux';
import messages from '../../Messages';
import searchFilter from '../../PresentationalComponents/Filters/SearchFilter';
import ErrorHandler from '../../PresentationalComponents/Snippets/ErrorHandler';
import { defaultReducers } from '../../store';
import { changeAffectedSystemsParams, clearAdvisorySystemsReducer,
    clearInventoryReducer, systemSelectAction } from '../../store/Actions/Actions';
import { inventoryEntitiesReducer, modifyAdvisorySystems } from '../../store/Reducers/InventoryEntitiesReducer';
import { exportAdvisorySystemsCSV, exportAdvisorySystemsJSON, fetchAdvisorySystems } from '../../Utilities/api';
import { remediationIdentifiers } from '../../Utilities/constants';
import {
    arrayFromObj, decodeQueryparams, persistantParams,
    remediationProvider, removeUndefinedObjectKeys
} from '../../Utilities/Helpers';
import {
    useBulkSelectConfig, useGetEntities, useOnExport, useRemoveFilter
} from '../../Utilities/Hooks';
import { intl } from '../../Utilities/IntlProvider';
import { advisorySystemsColumns, systemsRowActions } from '../Systems/SystemsListAssets';
import RemediationWizard from '../Remediation/RemediationWizard';
import AsyncRemediationButton from '../Remediation/AsyncRemediationButton';
import { useOnSelect, ID_API_ENDPOINTS } from '../../Utilities/useOnSelect';
import { systemsColumnsMerger, buildActiveFiltersConfig } from '../../Utilities/SystemsHelpers';
import advisoryStatusFilter from '../../PresentationalComponents/Filters/AdvisoryStatusFilter';
import { useSearchParams } from 'react-router-dom';

const AdvisorySystems = ({ advisoryName }) => {
    const dispatch = useDispatch();
    const store = useStore();
    const [searchParams, setSearchParams] = useSearchParams();

    const [isRemediationOpen, setRemediationOpen] = React.useState(false);
    const [
        RemediationModalCmp,
        setRemediationModalCmp
    ] = React.useState(() => () => null);

    const decodedParams = decodeQueryparams('?' + searchParams.toString());
    const systems = useSelector(({ entities }) => entities?.rows || [], shallowEqual);
    const status = useSelector(
        ({ entities }) => entities?.status || {}
    );
    const totalItems = useSelector(
        ({ entities }) => entities?.total || 0
    );
    const queryParams = useSelector(
        ({ AdvisorySystemsStore }) => AdvisorySystemsStore?.queryParams || {}
    );
    const selectedRows = useSelector(
        ({ entities }) => entities?.selectedRows || []
    );
    const metadata = useSelector(
        ({ AdvisorySystemsStore }) => AdvisorySystemsStore?.metadata || {}
    );

    const { systemProfile, selectedTags,
        filter, search, page, perPage, sort } = queryParams;

    React.useEffect(() => {
        apply(decodedParams);
        return () => {
            dispatch(clearInventoryReducer());
            dispatch(clearAdvisorySystemsReducer());
        };
    }, []);

    function apply(params) {
        dispatch(changeAffectedSystemsParams(params));
    }

    const [deleteFilters] = useRemoveFilter({ search, ...filter }, apply);

    const filterConfig = {
        items: [
            searchFilter(apply, search,
                intl.formatMessage(messages.labelsFiltersSystemsSearchTitle),
                intl.formatMessage(messages.labelsFiltersSystemsSearchPlaceholder)
            ),
            advisoryStatusFilter(apply, filter)
        ]
    };

    const activeFiltersConfig = buildActiveFiltersConfig(filter, search, deleteFilters);

    const showRemediationModal = useCallback(async (data) => {
        const resolvedData = await data;
        setRemediationModalCmp(() =>
            () => <RemediationWizard
                data={resolvedData}
                isRemediationOpen
                setRemediationOpen={setRemediationOpen} />);
        setRemediationOpen(!isRemediationOpen);
    }, [isRemediationOpen]);

    const onSelect = useOnSelect(
        systems,
        selectedRows,
        {
            endpoint: ID_API_ENDPOINTS.advisorySystems(advisoryName),
            queryParams,
            selectionDispatcher: systemSelectAction
        }
    );

    const selectedCount = selectedRows && arrayFromObj(selectedRows).length;

    const getEntites = useGetEntities(fetchAdvisorySystems, apply, { id: advisoryName }, setSearchParams);

    const onExport = useOnExport(advisoryName, queryParams, {
        csv: exportAdvisorySystemsCSV,
        json: exportAdvisorySystemsJSON
    }, dispatch);

    const remediationDataProvider = () => remediationProvider(
        advisoryName,
        removeUndefinedObjectKeys(selectedRows),
        remediationIdentifiers.advisory
    );

    const bulkSelectConfig = useBulkSelectConfig(
        selectedCount, onSelect, { total_items: totalItems }, systems, null, queryParams
    );

    return (
        <React.Fragment>
            {isRemediationOpen && <RemediationModalCmp /> || null}
            {(status.hasError || metadata?.has_systems === false)
                && <ErrorHandler code={status.code} metadata={metadata} />
                || <InventoryTable
                    isFullView
                    autoRefresh
                    initialLoading
                    ignoreRefresh
                    hideFilters={{ all: true, tags: false, operatingSystem: false }}
                    columns={(defaultColumns) => systemsColumnsMerger(defaultColumns, advisorySystemsColumns)}
                    showTags
                    customFilters={{
                        patchParams: {
                            search,
                            filter,
                            systemProfile,
                            selectedTags
                        }
                    }}
                    paginationProps={{
                        isDisabled: totalItems === 0
                    }}
                    onLoad={({ mergeWithEntities }) => {
                        store.replaceReducer(combineReducers({
                            ...defaultReducers,
                            ...mergeWithEntities(
                                inventoryEntitiesReducer(advisorySystemsColumns(false), modifyAdvisorySystems),
                                persistantParams({ page, perPage, sort, search }, decodedParams)
                            )
                        }));
                    }}
                    getEntities={getEntites}
                    tableProps={{
                        actionResolver: (row) => systemsRowActions(showRemediationModal, undefined, undefined, row),
                        canSelectAll: false,
                        variant: TableVariant.compact, className: 'patchCompactInventory', isStickyHeader: true
                    }}
                    filterConfig={filterConfig}
                    activeFiltersConfig={activeFiltersConfig}
                    exportConfig={{
                        isDisabled: totalItems === 0,
                        onSelect: onExport
                    }}
                    bulkSelect={onSelect && bulkSelectConfig}
                    dedicatedAction={(
                        <AsyncRemediationButton
                            remediationProvider={remediationDataProvider}
                            isDisabled={
                                arrayFromObj(selectedRows).length === 0
                            }
                        />
                    )}

                />
            }
        </React.Fragment>
    );
};

AdvisorySystems.propTypes = {
    advisoryName: propTypes.string
};

export default AdvisorySystems;
