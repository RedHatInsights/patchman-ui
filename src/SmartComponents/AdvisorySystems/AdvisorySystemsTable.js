import React from 'react';
import { TableVariant } from '@patternfly/react-table';
import { InventoryTable } from '@redhat-cloud-services/frontend-components/Inventory';
import propTypes from 'prop-types';
import { shallowEqual, useDispatch, useSelector, useStore } from 'react-redux';
import { combineReducers } from 'redux';
import messages from '../../Messages';
import searchFilter from '../../PresentationalComponents/Filters/SearchFilter';
import { defaultReducers } from '../../store';
import {
    systemSelectAction
} from '../../store/Actions/Actions';
import { inventoryEntitiesReducer, modifyAdvisorySystems } from '../../store/Reducers/InventoryEntitiesReducer';
import { exportAdvisorySystemsCSV, exportAdvisorySystemsJSON, fetchAdvisorySystems } from '../../Utilities/api';
import { remediationIdentifiers } from '../../Utilities/constants';
import {
    arrayFromObj, persistantParams,
    remediationProvider, removeUndefinedObjectKeys
} from '../../Utilities/Helpers';
import {
    useBulkSelectConfig, useGetEntities, useOnExport, useRemoveFilter, useOnSelect, ID_API_ENDPOINTS
} from '../../Utilities/hooks';
import { intl } from '../../Utilities/IntlProvider';
import { advisorySystemsColumns, systemsRowActions } from '../Systems/SystemsListAssets';
import AsyncRemediationButton from '../Remediation/AsyncRemediationButton';
import { systemsColumnsMerger, buildActiveFiltersConfig } from '../../Utilities/SystemsHelpers';
import advisoryStatusFilter from '../../PresentationalComponents/Filters/AdvisoryStatusFilter';

const AdvisorySystemsTable = ({
    advisoryName,
    apply,
    setSearchParams,
    activateRemediationModal,
    decodedParams
}) => {
    const dispatch = useDispatch();
    const store = useStore();

    const systems = useSelector(({ entities }) => entities?.rows || [], shallowEqual);
    const totalItems = useSelector(
        ({ entities }) => entities?.total || 0
    );
    const queryParams = useSelector(
        ({ AdvisorySystemsStore }) => AdvisorySystemsStore?.queryParams || {}
    );
    const selectedRows = useSelector(
        ({ entities }) => entities?.selectedRows || []
    );

    const { systemProfile, selectedTags,
        filter, search, page, perPage, sort } = queryParams;

    const [deleteFilters] = useRemoveFilter({ search, ...filter }, apply);

    const filterConfig = {
        items: [
            searchFilter(apply, search,
                intl.formatMessage(messages.labelsFiltersSystemsSearchTitle),
                intl.formatMessage(messages.labelsFiltersSystemsSearchPlaceholder)
            ),
            advisoryStatusFilter(apply, filter)
        ]
    };

    const activeFiltersConfig = buildActiveFiltersConfig(filter, search, deleteFilters);

    const onSelect = useOnSelect(
        systems,
        selectedRows,
        {
            endpoint: ID_API_ENDPOINTS.advisorySystems(advisoryName),
            queryParams,
            selectionDispatcher: systemSelectAction,
            totalItems
        }
    );

    const selectedCount = selectedRows && arrayFromObj(selectedRows).length;

    const getEntites = useGetEntities(fetchAdvisorySystems, apply, { id: advisoryName }, setSearchParams);

    const onExport = useOnExport(advisoryName, queryParams, {
        csv: exportAdvisorySystemsCSV,
        json: exportAdvisorySystemsJSON
    }, dispatch);

    const remediationDataProvider = () => remediationProvider(
        advisoryName,
        removeUndefinedObjectKeys(selectedRows),
        remediationIdentifiers.advisory
    );

    const bulkSelectConfig = useBulkSelectConfig(
        selectedCount, onSelect, { total_items: totalItems }, systems, null, queryParams
    );

    return (
        <InventoryTable
            isFullView
            autoRefresh
            initialLoading
            ignoreRefresh
            hideFilters={{ all: true, tags: false, operatingSystem: false }}
            columns={(defaultColumns) => systemsColumnsMerger(defaultColumns, advisorySystemsColumns)}
            showTags
            customFilters={{
                patchParams: {
                    search,
                    filter,
                    systemProfile,
                    selectedTags
                }
            }}
            paginationProps={{
                isDisabled: totalItems === 0
            }}
            onLoad={({ mergeWithEntities }) => {
                store.replaceReducer(combineReducers({
                    ...defaultReducers,
                    ...mergeWithEntities(
                        inventoryEntitiesReducer(advisorySystemsColumns(false), modifyAdvisorySystems),
                        persistantParams({ page, perPage, sort, search }, decodedParams)
                    )
                }));
            }}
            getEntities={getEntites}
            tableProps={{
                actionResolver: (row) => systemsRowActions(activateRemediationModal, undefined, undefined, row),
                canSelectAll: false,
                variant: TableVariant.compact, className: 'patchCompactInventory', isStickyHeader: true
            }}
            filterConfig={filterConfig}
            activeFiltersConfig={activeFiltersConfig}
            exportConfig={{
                isDisabled: totalItems === 0,
                onSelect: onExport
            }}
            bulkSelect={onSelect && bulkSelectConfig}
            dedicatedAction={(
                <AsyncRemediationButton
                    remediationProvider={remediationDataProvider}
                    isDisabled={
                        arrayFromObj(selectedRows).length === 0
                    }
                />
            )}

        />
    );
};

AdvisorySystemsTable.propTypes = {
    advisoryName: propTypes.string,
    apply: propTypes.func,
    setSearchParams: propTypes.func,
    activateRemediationModal: propTypes.func,
    decodedParams: propTypes.object
};

export default AdvisorySystemsTable;
