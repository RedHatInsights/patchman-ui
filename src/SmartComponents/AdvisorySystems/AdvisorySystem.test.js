/* eslint-disable no-unused-vars */
import toJson from 'enzyme-to-json';
import { act } from 'react-dom/test-utils';
import { Provider, useSelector } from 'react-redux';
import configureStore from 'redux-mock-store';
import { fetchIDs } from '../../Utilities/api';
import { systemRows } from '../../Utilities/RawDataForTesting';
import { initMocks } from '../../Utilities/unitTestingUtilities.js';
import AdvisorySystems from './AdvisorySystems';
import { BrowserRouter as Router } from 'react-router-dom';
import NoRegisteredSystems from '../../PresentationalComponents/Snippets/NoRegisteredSystems';

initMocks();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: jest.fn(() => ({ location: { search: 'testSearch' } }))
}));

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn()
}));

jest.mock('../../Utilities/api', () => ({
    ...jest.requireActual('../../Utilities/api'),
    fetchIDs: jest.fn(() => Promise.resolve({ ids: [] }).catch((err) => console.log(err)))
}));

jest.mock('@redhat-cloud-services/frontend-components/Inventory', () => ({
    ...jest.requireActual('@redhat-cloud-services/frontend-components/Inventory'),
    InventoryTable: jest.fn(() => <div className='testInventroyComponentChild'><div>This is child</div></div>)
}));

jest.mock(
    '../../PresentationalComponents/Filters/OsVersionFilter'
);

const mockState = {
    entities: {
        rows: systemRows,
        metadata: {
            limit: 25,
            offset: 0,
            total_items: 10
        },
        expandedRows: {},
        selectedRows: { 'test-system-1': true },
        error: {},
        status: 'resolved',
        total: 2
    },
    AdvisorySystemsStore: {
        queryParams: {}
    }
};

const initStore = () => {
    const customMiddleWare = () => next => action => {
        useSelector.mockImplementation(callback => {
            return callback(mockState);
        });
        next(action);
    };

    const mockStore = configureStore([customMiddleWare]);
    return mockStore(mockState);
};

let wrapper;
let store = initStore();
// eslint-disable-next-line no-unused-vars
const rejectedState = { entities: { ...mockState, status: 'rejected', error: { detail: 'test' } },
    AdvisorySystemsStore: mockState.AdvisorySystemsStore };
beforeEach(() => {
    useSelector.mockImplementation(callback => {
        return callback(mockState);
    });

    /**
 * This mount snapshot test is failing because of randomly generated key in PF component.
 * Before the snapshot was only shallow and did not display the real virtual DOM.
 * Shallow snapshot has the same result as the mount before.
 */
    wrapper = mount(<Provider store={store}>
        <Router> <AdvisorySystems advisoryName={'RHSA-2020:2755'} /></Router>
    </Provider>);
});

afterEach(() => {
    useSelector.mockClear();
    store.clearActions();
});

describe('AdvisorySystems.js', () => {

    it('Should match the snapshots and dispatch FETCH_AFFECTED_SYSTEMS only once', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Should display error page when status is rejected', () => {
        useSelector.mockImplementation(callback => {
            return callback({ AdvisorySystemsStore: rejectedState });
        });
        const testStore = initStore(rejectedState);
        const wrapper = mount(
            <Provider store={testStore}>
                <AdvisorySystems advisoryName = {'RHSA-2020:2755'} />
            </Provider>
        );
        expect(wrapper.find('Error')).toBeTruthy();
    });

    describe('test entity selecting', () => {
        it('Should unselect all', async() => {
            const { bulkSelect } = wrapper.find('.testInventroyComponentChild').parent().props();

            bulkSelect.items[0].onClick();
            const dispatchedActions = store.getActions();
            expect(dispatchedActions[1].type).toEqual('SELECT_ENTITY');
            expect(bulkSelect.items[0].title).toEqual('Select none (0)');
            expect(dispatchedActions[1].payload).toEqual([{ id: 'test-system-1', selected: false }]);
        });

        it('Should select a page', async () => {
            const { bulkSelect } = wrapper.find('.testInventroyComponentChild').parent().props();

            bulkSelect.items[1].onClick();
            const dispatchedActions = store.getActions();

            expect(dispatchedActions[1].type).toEqual('SELECT_ENTITY');
            expect(bulkSelect.items[1].title).toEqual('Select page (2)');
            expect(dispatchedActions[1].payload).toEqual([
                { id: 'test-system-id-1', selected: 'test-system-id-1' },
                { id: 'test-system-id-2', selected: 'test-system-id-2' }
            ]);
        });

        it('Should select all', async () => {
            const { bulkSelect } = wrapper.find('.testInventroyComponentChild').parent().props();

            bulkSelect.items[2].onClick();
            expect(fetchIDs).toHaveBeenCalledWith('/ids/advisories/RHSA-2020:2755/systems', { limit: -1 });
            expect(bulkSelect.items[2].title).toEqual('Select all (2)');
        });

        it('Should handle onSelect', async () => {
            const { bulkSelect } = wrapper.find('.testInventroyComponentChild').parent().props();

            bulkSelect.onSelect();
            const dispatchedActions = store.getActions();
            expect(dispatchedActions[1].type).toEqual('SELECT_ENTITY');
            expect(dispatchedActions[1].payload).toEqual([
                {
                    id: 'test-system-1',
                    selected: false
                }
            ]
            );
        });
    });

    it('Should open remediation modal', async () => {
        const { dedicatedAction } = wrapper.find('.testInventroyComponentChild').parent().props();
        expect(dedicatedAction).toMatchSnapshot();
    });

    it('Should clear store on unmount', () => {
        act(() => {
            wrapper.unmount();
        });
        const dispatchedActions = store.getActions();

        expect(dispatchedActions.filter(item => item.type === 'CLEAR_INVENTORY_REDUCER')).toHaveLength(1);
        expect(dispatchedActions.filter(item => item.type === 'CLEAR_ADVISORY_SYSTEMS_REDUCER')).toHaveLength(1);
    });

    it('Should display NoRegisteredSystems compnent if there are no systems registered', () => {
        const notFoundState = {
            ...mockState,
            status: 'rejected',
            error: {
                status: 403,
                title: 'testTitle',
                detail: 'testDescription'
            }
        };

        useSelector.mockImplementation(callback => {
            return callback({
                ...mockState,
                GlobalFilterStore: {},
                entities: {
                    ...mockState.entities,
                    metadata: { has_systems: false }
                }
            });
        });

        const tempStore = initStore(notFoundState);
        const tempWrapper = mount(<Provider store={tempStore}>
            <Router><AdvisorySystems /></Router>
        </Provider>);

        expect(tempWrapper.find(NoRegisteredSystems)).toBeTruthy();
    });
});
