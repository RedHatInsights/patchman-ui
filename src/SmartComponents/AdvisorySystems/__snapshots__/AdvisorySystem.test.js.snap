// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvisorySystems.js Should match the snapshots and dispatch FETCH_AFFECTED_SYSTEMS only once 1`] = `
<Provider
  store={
    {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      location={
        {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        }
      }
      navigationType="POP"
      navigator={
        {
          "action": "POP",
          "createHref": [Function],
          "createURL": [Function],
          "encodeLocation": [Function],
          "go": [Function],
          "listen": [Function],
          "location": {
            "hash": "",
            "key": "default",
            "pathname": "/",
            "search": "",
            "state": null,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
       
      <AdvisorySystems
        advisoryName="RHSA-2020:2755"
      >
        <mockConstructor
          activeFiltersConfig={
            {
              "deleteTitle": "Reset filters",
              "filters": [],
              "onDelete": [Function],
            }
          }
          autoRefresh={true}
          bulkSelect={
            {
              "checked": null,
              "isDisabled": false,
              "items": [
                {
                  "onClick": [Function],
                  "title": "Select none (0)",
                },
                {
                  "onClick": [Function],
                  "title": "Select page (2)",
                },
                {
                  "onClick": [Function],
                  "title": "Select all (2)",
                },
              ],
              "onSelect": [Function],
              "toggleProps": {
                "children": "     1 selected",
                "data-ouia-component-type": "bulk-select-toggle-button",
              },
            }
          }
          columns={[Function]}
          customFilters={
            {
              "patchParams": {
                "filter": undefined,
                "search": undefined,
                "selectedTags": undefined,
                "systemProfile": undefined,
              },
            }
          }
          dedicatedAction={
            <AsyncRemediationButton
              isDisabled={false}
              remediationProvider={[Function]}
            />
          }
          exportConfig={
            {
              "isDisabled": false,
              "onSelect": [Function],
            }
          }
          filterConfig={
            {
              "items": [
                {
                  "filterValues": {
                    "aria-label": "search-field",
                    "onChange": [Function],
                    "placeholder": "Filter by name",
                    "value": undefined,
                  },
                  "label": "System",
                  "type": "text",
                },
                {
                  "filterValues": {
                    "items": [
                      {
                        "label": "Installable",
                        "value": "Installable",
                      },
                      {
                        "label": "Applicable",
                        "value": "Applicable",
                      },
                    ],
                    "onChange": [Function],
                    "placeholder": "Filter by status",
                    "value": undefined,
                  },
                  "label": "Status",
                  "type": "checkbox",
                },
              ],
            }
          }
          getEntities={[Function]}
          hideFilters={
            {
              "all": true,
              "operatingSystem": false,
              "tags": false,
            }
          }
          ignoreRefresh={true}
          initialLoading={true}
          isFullView={true}
          onLoad={[Function]}
          paginationProps={
            {
              "isDisabled": false,
            }
          }
          showTags={true}
          tableProps={
            {
              "actionResolver": [Function],
              "canSelectAll": false,
              "className": "patchCompactInventory",
              "isStickyHeader": true,
              "variant": "compact",
            }
          }
        >
          <div
            className="testInventroyComponentChild"
          >
            <div>
              This is child
            </div>
          </div>
        </mockConstructor>
      </AdvisorySystems>
    </Router>
  </BrowserRouter>
</Provider>
`;

exports[`AdvisorySystems.js Should open remediation modal 1`] = `
<AsyncRemediationButton
  isDisabled={false}
  remediationProvider={[Function]}
/>
`;
